arangod/Aql/TraversalExecutor.cpp

```diff
@@ -0,0 +1,232 @@
+        return {_rowState, s};
+      }
+
+      if (!_input.isInitialized()) {
+        // We tried to fetch, but no upstream
+        TRI_ASSERT(_rowState == ExecutionState::DONE);
+        s.addFiltered(_traverser.getAndResetFilteredPaths());
+        s.addScannedIndex(_traverser.getAndResetReadDocuments());
+        return {_rowState, s};
+      }
+
+      resetTraverser();
+    }
+    if (!_traverser.hasMore() || !_traverser.next()) {
+      // Nothing more to read, reset input to refetch
+      _input = InputAqlItemRow{CreateInvalidInputRowHint{}};
+    } else {
+      // traverser now has next v, e, p values
+      if (_infos.useVertexOutput()) {
+        output.cloneValueInto(_infos.vertexRegister(), _input,
+                              _traverser.lastVertexToAqlValue());
+      }
+      if (_infos.useEdgeOutput()) {
+        output.cloneValueInto(_infos.edgeRegister(), _input,
+                              _traverser.lastEdgeToAqlValue());
+      }
+      if (_infos.usePathOutput()) {
+        transaction::BuilderLeaser tmp(_traverser.trx());
+        tmp->clear();
+        output.cloneValueInto(_infos.pathRegister(), _input,
+                              _traverser.pathToAqlValue(*tmp.builder()));
+      }
+      s.addFiltered(_traverser.getAndResetFilteredPaths());
+      s.addScannedIndex(_traverser.getAndResetReadDocuments());
+      return {computeState(), s};
+    }
+  }
+
+  s.addFiltered(_traverser.getAndResetFilteredPaths());
+  s.addScannedIndex(_traverser.getAndResetReadDocuments());
+  return {ExecutionState::DONE, s};
+}
+
+ExecutionState TraversalExecutor::computeState() const {
+  if (_rowState == ExecutionState::DONE && !_traverser.hasMore()) {
+    return ExecutionState::DONE;
+  }
+  return ExecutionState::HASMORE;
+}
+
+void TraversalExecutor::resetTraverser() {
+  // Initialize the Expressions within the options.
+  // We need to find the variable and read its value here. Everything is
+  // computed right now.
+  auto opts = _traverser.options();
+  opts->clearVariableValues();
+  for (auto const& pair : _infos.filterConditionVariables()) {
+    opts->setVariableValue(pair.first, _input.getValue(pair.second));
+  }
+
+  // Now reset the traverser
+  if (_infos.usesFixedSource()) {
+    auto pos = _infos.getFixedSource().find('/');
+    if (pos == std::string::npos) {
+      _traverser.options()->query()->registerWarning(
+          TRI_ERROR_BAD_PARAMETER,
+          "Invalid input for traversal: "
+          "Only id strings or objects with "
+          "_id are allowed");
+    } else {
+      // Use constant value
+      _traverser.setStartVertex(_infos.getFixedSource());
+    }
+  } else {
+    AqlValue const& in = _input.getValue(_infos.getInputRegister());
+    if (in.isObject()) {
+      try {
+        _traverser.setStartVertex(_traverser.options()->trx()->extractIdString(in.slice()));
+      } catch (...) {
+        // _id or _key not present... ignore this error and fall through
```

Comment:
Same here, fall through to where? What's with the start vertex now?

Reply:
good catch